elastic/detection-rules
# [Bug] [DaC] Metadata maturity field default mismatch and poor enforcement of rule naming conventions

# Pull Request

*Issue link(s)*:

Resolves https://github.com/elastic/detection-rules/issues/4282

## Summary - What I changed

In this PR I added explicit warnings when the rule file path does not match the expected path convention as if the rule was created via Kibana or the Detection Rules CLI rule prompt. This check occurs when the rules are added to a collection in the `kibana import-rules` and the `export-rules-from-repo` logic. The other routes assign the filename using the expected convention. This check simply compares the real path with what the constructed path would be if the rule was being written to the repo. If these do not match, it writes a warning message via click. 

Additionally, this PR addresses a bug where if a rule already exists in the repo, the maturity metadata field may change unexpectedly depending on the import path one uses. Originally, the default metadata value on importing rules to the repo was "development" if it came from an `.ndjson` file and `production` if it came from Kibana. In this way, rules could move from `production` back to `development` if an `.ndjson` file was used for the import instead of the Kibana CLI/API commands. This PR adds a check that if the file exists in the repo/on disk in the expected location, it will instead use the existing maturity value; otherwise, it will default to `development`

There is also an update that we may or may not want to include that is currently in this PR which, in addition to maturity, the creation and modified dates are now inherited from the file if they exist, and if not, the dates from Kibana are used. However, as this represents a potential change to how dates are calculated (before the date is specific to the date that the rule existed on disk in the repo) we may or may not want to include these changes and I would ask reviewers to leave their thoughts on this change in the PR. 

## How To Test

1. Create a rule in detection rules, not using the rule creation CLI where the file name does not match the naming convention and where the metadata created and modified dates are a date prior to the current day. Additionally, have a correctly named file where the metadata created and modified dates are a date prior to the current day and the maturity is set at `production`.
2. Import this rule into Kibana (or ndjson)
3. See the new warning printed to the screen
e.g.
```shell
❯ python -m detection_rules kibana --space demo --ignore-ssl-errors true import-rules
Loaded config file: /home/forteea1/Code/dac_demo/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

WARNING: Rule path does not match required path: defense_evasion_sts_get_federation_token.toml != defense_evasion_first_occurrence_of_sts_getfederationtoken_request_by_user.toml
WARNING: Rule path does not match required path: execution_suspicious_jar_child_process.toml != execution_deprecated_suspicious_java_child_process.toml
3 rule(s) successfully imported
 - 7a5cc9a8-5ea3-11ef-beec-f661ea17fbce
 - 8acb7614-1d92-4359-bfcf-478b6d9de150
 - 5930e5ed-37af-46ee-a77a-43db82af689c
```
5. Export this rule from Kibana (or ndjson)
6. See that the rule where the name matches that the maturity did not change.

## Checklist




## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)?
